### PR TITLE
Display full page when handling error or exception

### PR DESCRIPTION
**Pull request by Pierstoval**

### CODE REVIEW

File: src/Glpi/Application/ErrorHandler.php [hunk 1]
```diff
@@ -409,6 +413,7 @@ public function handleException(\Throwable $exception, bool $quiet = false): voi
 
         $this->logErrorMessage($error_type, $error_description, $error_trace, $log_level);
         if (!$quiet) {
+            $this->renderFullErrorPage($this->createFlattenedFromThrowable($exception));
             $this->outputDebugMessage($error_type, $error_description, $log_level);
```
[the remark this answers is not in the document]
Comment (cedric-anne) [reply src/Glpi/Application/ErrorHandler.php:417]: It is necessary in CLI context.

**Comment by Pierstoval** [comment]
| Code | Output |
| --- | --- |
| ```trigger_error('Error triggered', E_USER_ERROR); ``` | image  |
| ```throw new \RuntimeException('Exception triggered');``` | image |
| ```$a=[];for ($i=0;$i<100000;$i++){$a[]=$i;$a[]=$a;}``` | image |
| ```anything``` (parse error) | image |

File: src/Glpi/Application/ErrorHandler.php [hunk 2]
```diff
@@ -259,20 +264,20 @@ public function register(): void
...
 
-        if (in_array($error_code, self::FATAL_ERRORS)) {
-           // Fatal errors are handled by shutdown function
-           // (as some are not recoverable and cannot be handled here).
-           // Store backtrace to be able to use it there.
+        if (in_array($error_code, self::FATAL_ERRORS, true)) {
+            // Fatal errors are handled by shutdown function
+            // (as some are not recoverable and cannot be handled here).
+            // Store backtrace to be able to use it there.
             $this->last_fatal_trace = $error_trace;
```
Comment (cedric-anne) [inline src/Glpi/Application/ErrorHandler.php:278]: As fatal erros are now converted to exceptions, I guess there is no need to store this trace anymore, as fatal errors are now converted to exceptions.
```suggestion
```

File: src/Glpi/Application/ErrorHandler.php [hunk 3]
```diff
@@ -644,4 +638,40 @@ private function getTraceAsString(array $trace): string
...
+
+    private function convertErrorToThrowable(int $error_code, string $error_message, string $filename, int $line_number): \Throwable
+    {
+        $handler = new \Symfony\Component\ErrorHandler\ErrorHandler(null, $this->isDebug());
+        try {
+            $handler->handleError($error_code, $error_message, $filename, $line_number);
+        } catch (\Throwable $thrownException) {
+            return $thrownException;
+        }
+
```
[the remark this answers is not in the document]
Comment (Pierstoval) [reply src/Glpi/Application/ErrorHandler.php:649]: Symfony's ErrorHandler is good at converting PHP errors to Exceptions, the goal here is to call it and get the exception, but in some rare cases, `handleError()` will return `false`, therefore we create a custom exception that's "lower quality", but we can still have a record of it.

File: src/Glpi/Application/ErrorHandler.php [hunk 4]
```diff
@@ -393,23 +396,11 @@ public function handleException(\Throwable $exception, bool $quiet = false): voi
...
-
-        $log_level = self::ERROR_LEVEL_MAP[E_ERROR];
+        $this->logException($exception);
 
-        $this->logErrorMessage($error_type, $error_description, $error_trace, $log_level);
-        if (!$quiet) {
-            $this->outputDebugMessage($error_type, $error_description, $log_level);
+        if (!$quiet && isCommandLine()) {
+            $log_level = self::ERROR_LEVEL_MAP[$exception->getCode()] ?? 255;
+            $this->outputDebugMessage($exception->getCode(), $exception->getMessage(), $log_level);
```
Comment (cedric-anne) [inline src/Glpi/Application/ErrorHandler.php:403]: The full stack trace will not be present anymore in logs. This change should be reverted.

Comment (Pierstoval) [reply src/Glpi/Application/ErrorHandler.php:403]: I confirm that the stack trace is available in all possible cases of errors (I tested them) except for memory limit where we only have the file and line

Comment (cedric-anne) [reply src/Glpi/Application/ErrorHandler.php:403]: My bad, I misread the change. I mixed output and logs in my mind.

Comment (cedric-anne) [reply src/Glpi/Application/ErrorHandler.php:403]: But the previous output was more precise as it was containing the exception class, the file and the line..

Comment (Pierstoval) [reply src/Glpi/Application/ErrorHandler.php:403]: Here's an example of the logs when an exception is triggered:

```
[2024-08-01 13:58:31] glpiphplog.CRITICAL:   *** Uncaught Exception RuntimeException: Exception triggered in /var/www/glpi/front/error_exception.php at line 35
  Backtrace :
  ...Glpi/Controller/LegacyFileLoadController.php:57 require()
  vendor/symfony/http-kernel/HttpKernel.php:101      Glpi\Controller\LegacyFileLoadController->Glpi\Controller\{closure}()
  ...ymfony/http-foundation/StreamedResponse.php:106 Symfony\Component\HttpKernel\HttpKernel::Symfony\Component\HttpKernel\{closure}()
  vendor/symfony/http-foundation/Response.php:423    Symfony\Component\HttpFoundation\StreamedResponse->sendContent()
  public/index.php:58                                Symfony\Component\HttpFoundation\Response->send()
```

Comment (cedric-anne) [reply src/Glpi/Application/ErrorHandler.php:391]: The logs seems OK, but the output message that will be displayed in the console should not have changed. The text was more precise.
The `testHandleException` should not have changed too.

Comment (Pierstoval) [reply src/Glpi/Application/ErrorHandler.php:391]: The screenshots above are for the "non-dev" setups.

In dev, it provides the stack trace like this: 
image

Comment (cedric-anne) [reply src/Glpi/Application/ErrorHandler.php:391]: I am talking about the console output that changed, according to this change in tests:
image

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**